base/acme/src/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -539,4 +565,132 @@ public void updateOrder(ACMEAccount account, ACMEOrder order) throws Exception {
+        Set<String> authorizedDNSNames = new HashSet<>();
+
+        for (URI authzURL : authorizations) {
+
+            String authzPath = authzURL.getPath();
+            String authzID = authzPath.substring(authzPath.lastIndexOf('/') + 1);
+            ACMEAuthorization authz = database.getAuthorization(authzID);
+
+            // authz is guaranteed to be valid at this point
+
+            ACMEIdentifier identifier = authz.getIdentifier();
+            String type = identifier.getType();
+            String value = identifier.getValue();
+
+            // TODO: support other identifier types
+
+            if ("dns".equals(type)) {
+                // store normalized authorized DNS names
+                authorizedDNSNames.add(value.toLowerCase());
+            }
+        }
+
+        logger.info("Authorized DNS names:");
+        for (String dnsName : authorizedDNSNames) {
+            logger.info("- " + dnsName);
+        }
+
+        logger.info("Parsing CSR");
+        String strCSR = CryptoUtil.normalizeCertAndReq(csr);
+        byte[] binCSR = Utils.base64decode(strCSR);
+        PKCS10 pkcs10 = new PKCS10(binCSR);
+
+        // store normalized DNS names from CSR
+        Set<String> dnsNames = new HashSet<>();
+
+        X500Name subjectDN = pkcs10.getSubjectName();
+        logger.info("Parsing subject DN: " + subjectDN);
+
+        String cn;
+        try {
+            cn = subjectDN.getCommonName();
+
+        } catch (NullPointerException e) {
+            // X500Name.getCommonName() throws NPE if subject DN is blank
+            // TODO: fix X500Name.getCommonName() to return null
+            cn = null;
+        }
+
+        if (cn != null) {
+            dnsNames.add(cn.toLowerCase());
+        }
+
+        logger.info("Parsing CSR Attributes:");
+        PKCS10Attributes attributes = pkcs10.getAttributes();
+        for (PKCS10Attribute attribute : attributes) {
+
+            ObjectIdentifier attrID = attribute.getAttributeId();
+            CertAttrSet attrValues = attribute.getAttributeValue();
+            String attrName = attrValues.getName();
+            logger.info("- " + attrID + ": " + attrName);
+
+            // TODO: support other attributes
+
+            if (attrValues instanceof Extensions) {
+                Extensions extensions = (Extensions) attrValues;
+                parseCSRExtensions(dnsNames, extensions);
+            }
+        }
+
+        logger.info("Validating DNS names in CSR");
+        for (String dnsName : dnsNames) {
+            logger.info("- " + dnsName);
+        }
+
+        Set<String> unauthorizedDNSNames = new HashSet<>(dnsNames);
+        unauthorizedDNSNames.removeAll(authorizedDNSNames);
+
+        if (!unauthorizedDNSNames.isEmpty()) {
+            // TODO: generate proper exception
+            throw new Exception("Unauthorized DNS names: " + StringUtils.join(unauthorizedDNSNames, ", "));
+        }
+
+        // TODO: validate other things in CSR
```

Comment:
This is fine for now. But as this gets longer, the code will be less readable. Perhaps separate getters for DNS entries/other properties and a separate validation functions.

Reply:
Other things to validate:

 - No non-DNS entries
 - Key size is valid (2048+ if RSA).

Reply:
You mean the CSR for ACME cannot contain non-DNS entries?
Also does it mean the only supported key alg is RSA?

Reply:
Or should some of these validations done by the CA (or the profile)?

Reply:
@edewata [said](https://github.com/dogtagpki/pki/pull/285#discussion_r345995185):

>  Also does it mean the only supported key alg is RSA?

No, sorry. I meant that if the key is RSA, validate the keysize. EC keys don't really have the keysize problem.

---

@edewata [said](https://github.com/dogtagpki/pki/pull/285#discussion_r345996669):

> Or should some of these validations done by the CA (or the profile)?

Yeah that'd be best. But, I don't see here where the profile is being applied or checked? I guess, to me, profile applies when we know we're going to be issuing the certificate. We need/want policy before then. But perhaps I'm using the wrong terminology.

--- 

@edewata [said](https://github.com/dogtagpki/pki/pull/285#discussion_r345995185):

> You mean the CSR for ACME cannot contain non-DNS entries?

I don't see how it could, at this stage in ACME's life. See the [RFC 8555](https://tools.ietf.org/html/rfc8555#page-47):

>    The CSR encodes the client's requests with regard to the content of
>    the certificate to be issued.  The CSR MUST indicate the exact same
>    set of requested identifiers as the initial newOrder request.
>    Identifiers of type "dns" MUST appear either in the commonName
>    portion of the requested subject name or in an extensionRequest
>    attribute [RFC2985] requesting a subjectAltName extension, or both.
>    (These identifiers may appear in any sort order.)  Specifications
>    that define new identifier types must specify where in the
>    certificate signing request these identifiers can appear.

All identifiers must be in the new order request (so other identifiers not in the initial order request are not allowed). The only identifiers we can currently process are DNS identifiers [according to the RFC](https://tools.ietf.org/html/rfc8555#section-9.7.7).

So we need to be actively excluding non-DNS entries, at least as long as DNS-based challenges are the only types we have available (DNS-01 / HTTPS-01 / &c). 

My 2c.

Reply:
The profile name is specified here:
https://github.com/edewata/pki/blob/acme/base/acme/conf/backend/pki/backend.json#L5

In a later patch the profile name will be used to get the request template:
https://github.com/edewata/pki/blob/acme/base/acme/src/org/dogtagpki/acme/backend/PKIBackend.java#L126

The profile contains policies that will validate the CSR before the CA will issue the cert, so I think here we should only validate things that are relevant to ACME, and leave enrollment policies to the CA/profile.